numpy/typing/__init__.py

```diff
@@ -0,0 +1,65 @@
+"""
+============================
+Typing (:mod:`numpy.typing`)
+============================
+
+.. warning::
+
+  Some of the types in this module rely on features only present in
+  the standard library in Python 3.8 and greater. If you want to use
+  these types in earlier versions of Python, you should install the
+  typing-extensions_ package.
+
+Large parts of the NumPy API have PEP-484-style type annotations. In
+addition, the following type aliases are available for users.
+
+- ``typing.ArrayLike``: objects that can be converted to arrays
+- ``typing.DtypeLike``: objects that can be converted to dtypes
+
+Roughly speaking, ``typing.ArrayLike`` is "objects that can be used as
+inputs to ``np.array``" and ``typing.DtypeLike`` is "objects that can
+be used as inputs to ``np.dtype``".
+
+.. _typing-extensions: https://pypi.org/project/typing-extensions/
+
+Differences from the runtime NumPy API
+--------------------------------------
+
+NumPy is very flexible. Trying to describe the full range of
+possibilities statically would result in types that are not very
+helpful. For that reason, the typed NumPy API is often stricter than
+the runtime NumPy API. This section describes some notable
+differences.
+
+ArrayLike
+~~~~~~~~~
+
+The ``ArrayLike`` type tries to avoid creating object arrays. For
+example,
+
+.. code-block:: python
+
+  np.array(x**2 for x in range(10))
+
+is valid NumPy code which will create an object array. The types will
+complain about this usage however.
+
```

Comment:
Can we add a brief note on the suggested work-around?

The obvious way would be to add a comment disabling typing:
```
  np.array(x**2 for x in range(10))  # type: ignore
```

Are there other recommended options?

I think we've also discussed making checks less strict if `dtype=object` is specified, e.g.,
```
  np.array(x**2 for x in range(10), dtype=object)
```
I don't know if that works yet. If it does, perhaps we should mention it, too.

Reply:
> Are there other recommended options?

The other way we test for:

https://github.com/numpy/numpy/blob/master/numpy/tests/typing/pass/array_like.py#L43)

is adding an explicit `Any` annotation. I've added examples of both methods to the docs.

> I don't know if that works yet. If it does, perhaps we should mention it, too.

Seems like @seberg would probably know the answer to that?